model_evaluation/utils_export/dataset.py

```diff
@@ -0,0 +1,270 @@
+    """
+
+    utils_cloudml.is_valid_spec(feature_keys_spec)
+    if example_key in feature_keys_spec:
+      raise ValueError('example_key should not be part of input_data.'
+                       'It will be created when writing to tf-records')
+
+    self._model_name = model_names
+    self._feature_keys_spec = feature_keys_spec
+    self._prediction_keys = prediction_keys
+    self._project_name = project_name
+    self._example_key = example_key
+
+  def feature_keys_spec(self):
+    return self._feature_keys_spec
+
+  def example_key(self):
+    return self._example_key
+
+  def model_names(self):
+    return self._model_name
+
+  def prediction_keys(self):
+    return self._prediction_keys
+
+  def project_name(self):
+    return self._project_name
+
+  def set_job_ids_prediction(self, job_ids):
+    self._job_ids_prediction = job_ids
+
+  def job_ids_prediction(self):
+    return self._job_ids_prediction
+
+
+class Dataset(object):
+  """Defines a format for every dataset to work with evaluation pipeline.
+
+  Usage:
+
+  input_fn = ... (returns pandas DataFrame).
+  dataset = Dataset(input_fn) # Verifies that input_fn is ok.
+
+  dataset.loads_data(10000, 0.5)
+
+  model = Model(...)
+  # Next function verifies that models are compatible.
+  dataset.add_model_prediction_to_data(model)
+
+  dataset.show_data()
+  """
+
+  def __init__(self, input_fn):
+    self.check_input_fn(input_fn)
+    self._input_fn = input_fn
+
+  def show_data(self):
+    if not hasattr(self, 'data'):
+      raise ValueError(
+          'Dataset does not have data yet.'
+          ' You need to run `loads_data` first.')
+    return self.data
+
+  def check_input_fn(self, input_fn):
+    """Checks if the input_fn meets requirements."""
+    args_input_fn = inspect.getargspec(input_fn).args
+    if 'max_n_examples' not in args_input_fn:
+      raise ValueError(
+          'input_fn should have (at least) `max_n_examples`'
+          ' as arguments.')
+
+    loaded_data = input_fn(max_n_examples=1)
+
+    if not isinstance(loaded_data, pd.DataFrame):
+      raise ValueError('input_fn should return a pandas DataFrame.')
+
+    if len(loaded_data) != 1:
+      raise ValueError(
+          'input_fn(max_n_examples=1) should contain 1 row (exactly).')
+    logging.info('input_fn is compatible with the `Dataset` class.')
+
+  def check_compatibility(self, model):
+    """Checks that input_fn is compatible with the model."""
+
+    if hasattr(self, 'data'):
+      test_df = self.data
+    else:
+      test_df = self._input_fn(max_n_examples=1)
+
+    for key in model.feature_keys_spec():
+      if key not in test_df.columns:
+        raise ValueError(
+            'input_fn must contain at least the feature keys {}'.format(
+                model.feature_keys_spec()))
+    logging.info('Model is compatible with the `Dataset` instance.')
+
+  def load_data(self, max_n_examples, **kwargs):
+    self.data = self._input_fn(max_n_examples, **kwargs)
```

Comment:
Maybe should be 'max_n_examples = max_n_examples' since we can't guarantee that max_n_examples is the first argument of the input function?